sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala

```diff
@@ -959,6 +959,37 @@ class CollationStringExpressionsSuite
     assert(collationMismatch.getErrorClass === "DATATYPE_MISMATCH.UNEXPECTED_INPUT_TYPE")
   }
 
+  test("DateFormat string expression with collation") {
+    case class DateFormatTestCase[R](d: String, f: String, c: String, r: R)
+    val testCases = Seq(
+      DateFormatTestCase("2021-01-01", "yyyy-MM-dd", "UTF8_BINARY", "2021-01-01"),
+      DateFormatTestCase("2021-01-01", "yyyy-dd", "UTF8_BINARY_LCASE", "2021-01"),
+      DateFormatTestCase("2021-01-01", "yyyy-MM-dd", "UNICODE", "2021-01-01"),
+      DateFormatTestCase("2021-01-01", "yyyy", "UNICODE_CI", "2021")
+    )
+
+    testCases.foreach(t => {
+      val query = s"SELECT date_format(collate('${t.d}', '${t.c}'), '${t.f}')"
+      // Result & data type
+      checkAnswer(sql(query), Row(t.r))
+      assert(sql(query).schema.fields.head.dataType.sameType(SQLConf.get.defaultStringType))
```

Comment:
Please consider explicitly varying default collation config such as in https://github.com/apache/spark/blob/b14abb3a2ed086d2ff8f340f60c0dc1e460c7a59/sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala#L442-L446

Reply:
+1, this is the way to go (you'll see this once you switch over to `CollationSQLExpressionsSuite` too)

Reply:
Found a bug which is blocking collate expression to be executed when there is a default collation config set. Will sync with @mihailom-db to figure out path forward.

Reply:
Done.